### DISCUSSION TITLE
ESM vs UMD Performance

### DISCUSSION THREAD

**Discussion by jasonsturges** [pixijs/pixijs#8462] [General]
Tracing some performance issues, think I'm seeing that ESM modules are 3x slower:

ESM: animation frame takes ~12ms averaging 80fps
<img width="2056" alt="Screen Shot 2022-07-01 at 12 57 19 AM" src="https://user-images.githubusercontent.com/1213591/176833701-5d838acf-f36a-434e-b7fb-eac687126515.png">

Browser UMD: animation frame takes ~4ms averaging 120fps
<img width="2056" alt="Screen Shot 2022-07-01 at 12 58 24 AM" src="https://user-images.githubusercontent.com/1213591/176833794-db9a4111-0833-46ac-8a44-ec6680ab01cb.png">

Same application, view, state... all using default PIXI.Application Ticker.

Just initial discovery here, but any immediate thoughts regarding this performance?

Pixi.js: v6.3.2
Chrome: Version 103.0.5060.53 (Official Build) (arm64)
MacBook Pro 16", Apple M1 Pro

**Comment by ivanpopelyshev** [comment]
Long Live UMD!!11!!!11!!!

**Reply by ivanpopelyshev** [reply]
I cant find forum topics and issues, but certainly this appeared multiple times. People usually solved it with changing their babel configuration

**Comment by bigtimebuddy** [comment]
Interesting finding. Would love to profile this. Do you have something that we can easily compare? Specifically interested in how it's being built too.

We don't have UMD, so are you talking about CommonJS vs ESM or the IIFE browser bundle vs MJS?

**Comment by jasonsturges** [comment]
@ivanpopelyshev I think you're right... yeah, my configuration is likely suboptimal.

@bigtimebuddy Difficult to fully showcase this, but I might email you some links.

However, my [pixi-experiments](https://github.com/jasonsturges/pixi-experiments) exhibits similar behavior of the problem side from the Performance -> Stress Test story:
https://jasonsturges.github.io/pixi-experiments/?path=/story/performance-stress-test--bursts

Story code: [stories/performance/StressTest.stories.js](https://github.com/jasonsturges/pixi-experiments/blob/main/stories/performance/StressTest.stories.js)

Run dev with `npm run storybook`

Scenario that I'm seeing this:
- TypeScript library uses Rollup defining Pixi.js as external - basically [typescript-npm-package](https://github.com/jasonsturges/typescript-npm-package)
- Storybook uses that library for testing, building with Webpack 4 - basically [storybook-for-pixi.js](https://github.com/jasonsturges/storybook-for-pixi.js)
- Create React App uses that library in production

Storybook ESM is considerably slower than the production app with CommonJS (sorry, was wrong here).  Probably something else in play besides modules, but wanted to sanity check this.

Thanks for the insight - I'll circle back after I obtain something more tangible for discussion.

**Comment by Julien-Marcou** [comment]
Have you tried using the `next` branch of PixiJS?

I'm wondering if this performance penalty could come from the downleveling to `es5` (which is no longer the case on the `next` branch)

**Reply by jasonsturges** [reply]
@Julien-Marcou  Interesting - that's a great idea, I'll try that.
